src/VisualStudio/Core/Def/Packaging/PackageInstallerServiceFactory.cs

```diff
@@ -396,171 +399,123 @@ private void OnWorkspaceChanged(object sender, WorkspaceChangeEventArgs e)
-
-                            ProcessBatchedChangesOnForeground(cancellationToken);
-                        },
-                        cancellationToken,
-                        TaskContinuationOptions.OnlyOnRanToCompletion | TaskContinuationOptions.ExecuteSynchronously,
-                        TaskScheduler.Default).Unwrap();
-            }
+            Contract.ThrowIfNull(_workQueue, "We should only register for events after having create the WorkQueue");
+            _workQueue.AddWork((solutionChanged, changedProject));
         }
 
-        private void ProcessBatchedChangesOnForeground(CancellationToken cancellationToken)
+        private async Task ProcessWorkQueueAsync(
+            ImmutableArray<(bool solutionChanged, ProjectId? changedProject)> workQueue, CancellationToken cancellationToken)
         {
-            this.AssertIsForeground();
+            ThisCanBeCalledOnAnyThread();
 
-            // If we've been asked to stop, then there's no point proceeding.
-            if (cancellationToken.IsCancellationRequested)
-            {
-                return;
-            }
+            Contract.ThrowIfNull(_workQueue, "How could we be processing a workqueue change without a workqueue?");
 
             // If we've been disconnected, then there's no point proceeding.
             if (_workspace == null || !IsEnabled)
-            {
                 return;
-            }
 
-            // Get a project to process.
-            var solution = _workspace.CurrentSolution;
-            var projectId = DequeueNextProject(solution);
-            if (projectId == null)
-            {
-                // No project to process, nothing to do.
+            var serviceContainer = (IBrokeredServiceContainer)await _asyncServiceProvider.GetServiceAsync(typeof(SVsBrokeredServiceContainer)).ConfigureAwait(false);
+            var serviceBroker = serviceContainer?.GetFullAccessServiceBroker();
+            if (serviceBroker == null)
                 return;
-            }
 
-            // Process this single project.
-            ProcessProjectChange(solution, projectId);
+            var nugetService = await serviceBroker.GetProxyAsync<INuGetProjectService>(NuGetServices.NuGetProjectServiceV1, cancellationToken: cancellationToken).ConfigureAwait(false);
 
-            // After processing this single project, yield so the foreground thread
-            // can do more work.  Then go and loop again so we can process the 
-            // rest of the projects.
-            Task.Factory.SafeStartNewFromAsync(
-                async () =>
-                {
-                    await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+            using (nugetService as IDisposable)
+            {
+                // If we didn't get a nuget service, there's nothing we can do in terms of querying the solution for
+                // nuget info.
+                if (nugetService == null)
+                    return;
```

Comment:
Write this before the using? It looks a bit funky otherwise to be doing the using (after a cast) and only then ever checking if it's null.

Reply:
We have an analyzer to ensure people dispose of the returned proxy that is pretty narrow when it comes to allowed patterns. The if check within the using is one of those quirky requirements.

Reply:
yeah, sorry, appeasing the analyzing here :)